### PR TITLE
Fix and add the documentation for Built-in Serializers and Deserializers

### PR DESCRIPTION
**Pull request by Lalitha333**
### Checklist

- [ ] New code follows the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html)
- [ ] If necessary, new public API validates arguments, for example rejects `null`
- [ ] New public API has Javadoc
    - [ ] Javadoc uses `@since $next-version$`  
      (`$next-version$` is a special placeholder which is automatically replaced during release)
- [ ] If necessary, new unit tests have been added  
  - [ ] Assertions in unit tests use [Truth](https://truth.dev/), see existing tests
  - [ ] No JUnit 3 features are used (such as extending class `TestCase`)
  - [ ] If this pull request fixes a bug, a new test was added for a situation which failed previously and is now fixed
- [ ] `mvn clean verify javadoc:jar` passes without errors

### CODE REVIEW

File: UserGuide.md [hunk 1]
```diff
@@ -378,10 +378,64 @@ This approach is practical only if the array appears as a top-level element or i
...
+    * java.sql.Date: Serialized as a string in the format "yyyy-MM-dd".
+    * java.sql.Time: Serialized as a string in the format "HH:mm:ss".
+    * java.sql.Timestamp: Serialized as a string in the format "yyyy-MM-dd HH:mm:ss".
+5. BigInteger and BigDecimal:
+    * java.math.BigInteger: Serialized as a string.
+    * java.math.BigDecimal: Serialized as a string.
+6. Enumerations:
+    * Enum types are serialized as strings by default. You can also customize the serialization using @SerializedName annotation.
+7. Optional:
+    * java.util.Optional: Serializes the value if present; otherwise, it serializes as null.
```
Comment (Marcono1234) [inline UserGuide.md:434]: This is wrong? See #1102

File: UserGuide.md [hunk 2]
```diff
@@ -378,10 +378,64 @@ This approach is practical only if the array appears as a top-level element or i
...
+8. Custom Objects:
+    * Gson can automatically serialize and deserialize custom Java objects using reflection. It serializes the non-static, non-transient fields of an object.
```
Comment (Marcono1234) [inline UserGuide.md:436]:
```suggestion
    * Gson can automatically serialize and deserialize custom Java objects using reflection. By default, it serializes the non-static, non-transient fields of an object.
```
(Because this can be customized)

File: UserGuide.md [hunk 3]
```diff
@@ -378,10 +378,64 @@ This approach is practical only if the array appears as a top-level element or i
...
+    * java.math.BigInteger: Serialized as a string.
+    * java.math.BigDecimal: Serialized as a string.
+6. Enumerations:
+    * Enum types are serialized as strings by default. You can also customize the serialization using @SerializedName annotation.
+7. Optional:
+    * java.util.Optional: Serializes the value if present; otherwise, it serializes as null.
+8. Custom Objects:
+    * Gson can automatically serialize and deserialize custom Java objects using reflection. It serializes the non-static, non-transient fields of an object. 
+    
+It's important to note that Gson's built-in serializers and deserializers can handle nested objects and collections of objects as well. If a class is not supported by Gson's default behavior, you can provide custom serialization and deserialization logic using Gson's JsonSerializer and JsonDeserializer interfaces.
```
Comment (Marcono1234) [inline UserGuide.md:438]: Maybe this should rather recommend `TypeAdapter` since the documentation for `JsonSerializer` and `JsonDeserializer` also recommend `TypeAdapter` for performance reasons.
```suggestion
It's important to note that Gson's built-in serializers and deserializers can handle nested objects and collections of objects as well. If a class is not supported by Gson's default behavior, you can provide custom serialization and deserialization logic using Gson's `TypeAdapter` class.
```

Or maybe even omit this sentence because the section "Custom Serialization and Deserialization" is directly below this.

File: UserGuide.md [hunk 4]
```diff
@@ -378,10 +378,64 @@ This approach is practical only if the array appears as a top-level element or i
...
+1. Primitive Types:
+    * `int, Integer, long, Long, float, Float, double, Double, boolean, Boolean, char, Character, byte, Byte, short, Short.`
+2. Arrays:
+    * Arrays of primitive types and their wrappers.
+    * Arrays of objects.
+3. Collections:
+    * java.util.List: List of objects.
+    * java.util.Set: Set of objects.
+    * java.util.Map: Map of key-value pairs.
+4. Date and Time:
```
Comment (Marcono1234) [inline UserGuide.md:423]: Maybe this should also refer to `GsonBuilder.setDateFormat(...)`

File: UserGuide.md [hunk 5]
```diff
@@ -378,10 +378,64 @@ This approach is practical only if the array appears as a top-level element or i
 
 ### Built-in Serializers and Deserializers
 
-Gson has built-in serializers and deserializers for commonly used classes whose default representation may be inappropriate, for instance
+Gson has built-in serializers and deserializers for commonly used classes whose default representation may be inappropriate.
+Gson provides built-in serializers and deserializers for basic Java types (e.g., primitive types, strings, arrays) as well as some commonly used classes like `Date`, `BigInteger`, `BigDecimal`, and more. When you use Gson to convert Java objects to JSON or JSON to Java objects, these built-in serializers and deserializers handle the conversion automatically for these basic types and classes.
 
-* `java.net.URL` to match it with strings like `"https://github.com/google/gson/"`
-* `java.net.URI` to match it with strings like `"/google/gson/"`
```
Comment (Marcono1234) [inline UserGuide.md:384]: Maybe it would be good to include `URL` and `URI` in the list below again with a description like "Serialized as their string representation" or similar, since these types might also be somewhat common.

File: UserGuide.md [hunk 6]
```diff
@@ -378,10 +378,64 @@ This approach is practical only if the array appears as a top-level element or i
...
+6. Enumerations:
+    * Enum types are serialized as strings by default. You can also customize the serialization using @SerializedName annotation.
+7. Optional:
+    * java.util.Optional: Serializes the value if present; otherwise, it serializes as null.
+8. Custom Objects:
+    * Gson can automatically serialize and deserialize custom Java objects using reflection. It serializes the non-static, non-transient fields of an object. 
+    
+It's important to note that Gson's built-in serializers and deserializers can handle nested objects and collections of objects as well. If a class is not supported by Gson's default behavior, you can provide custom serialization and deserialization logic using Gson's JsonSerializer and JsonDeserializer interfaces.
 
 For many more, see the internal class [`TypeAdapters`](gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java).
```
Comment (Marcono1234) [inline UserGuide.md:440]: Maybe reword this a bit to give a bit more context, for example:
```suggestion
For more of the built-in serializers and deserializers, see the internal class [`TypeAdapters`](gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java).
```

**Review by Marcono1234 — COMMENTED** [review]
Thanks! This looks really useful, and is definitely better than only referring the user to the internal implementation details..

If you think this resolves the GitHub issue you mentioned in the description, could you please edit the description to contain "Closes #..."; this way [GitHub automatically recognizes this](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests).

A few of my comments below might be nit-picky, feel free to consider them as suggestions, you don't have to follow them strictly. However, it looks like in some cases you also described the serialized output incorrectly.

Note that I am not a direct project member, and they might think differently.

File: UserGuide.md [hunk 7]
```diff
@@ -378,10 +378,64 @@ This approach is practical only if the array appears as a top-level element or i
...
+```java
+String json = "{\"name\":\"Jane Smith\",\"age\":25}";
+Person person = gson.fromJson(json, Person.class);
+```
+The `gson.fromJson()` method uses the built-in deserializers to convert the JSON back into a `Person` object.
+
+Gson also allows you to customize serialization and deserialization by providing your own custom serializers and deserializers for specific types if needed.
+
+Below is a list of some of the classes supported by Gson's built-in serializers and deserializers:
+
```
[the remark this answers is not in the document]
Comment (Lalitha333) [reply UserGuide.md:413]: @Marcono1234 Sure, changed to unordered list.

[the remark this answers is not in the document]
Comment (Marcono1234) [reply UserGuide.md:413]: @Lalitha333, just to clarify, you are still working on these changes, right? Or did you forget to push the commits?

[the remark this answers is not in the document]
Comment (Lalitha333) [reply UserGuide.md:413]: Sorry for the delay, yes I’m going to push the changes as per review comments.

**Comment by Marcono1234** [comment]
Thanks for your changes! Do you want to address some or all of the other review comments as well?

**Comment by Lalitha333** [comment]
> Thanks for your changes! Do you want to address some or all of the other review comments as well?

I believe I made all the requested changes. Please let me know if I have missed any.

File: UserGuide.md [hunk 8]
```diff
@@ -378,12 +378,67 @@ This approach is practical only if the array appears as a top-level element or i
...
+* Primitive Types:
+    * `int`, `long`, `float`, `double`, `boolean`, `char`, `byte`, `short` and their wrapper types.
+* Arrays:
+    * Arrays of primitive types and their wrappers.
+    * Arrays of objects.
+* Collections:
+    * `java.util.Collection`: Collection of objects, and collection subtypes such as:
+      * `java.util.List`: List of objects.
+      * `java.util.Set`: Set of objects.
+      * `java.util.Map`: Map of key-value pairs.
```
Comment (Marcono1234) [inline UserGuide.md:423]: Should probably unindent this by one level because `Map` is not a subtype of `Collection`
```suggestion
    * `java.util.Map`: Map of key-value pairs.
```

**Comment by Marcono1234** [comment]
@Lalitha333, yes there are a few unresolved points (I also just added a new single comment about your follow-up commit); see the [GitHub documentation](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#resolving-conversations) for how you can browse them.

If you addressed a comment, feel free to mark it as "resolved". Also, in case you don't want to address a comment, it might be good if you could write a short reply explaining this, so it is clear that you did not accidentally marked it as "resolved".
